### ISSUE TITLE
schema/gen/go: produce a single Go file

### ISSUE THREAD

**Issue by mvdan** [ipld/go-ipld-prime#93]
Right now, we generate a separate Go file for each schema type, plus a few extra files like `minima.go` and `typeTable.go`. We could really do with far fewer files, even a single one for the entire generated code.

Here are reasons why a linear amount of files isn't a good idea:

* Removing a schema type and re-running `go generate` leaves orphaned files behind.
* Directories get cluttered, making Go development harder. It even pushes me to generate code in a separate internal package.
* Many of the generated files are tiny; in my current HAMT ADL code, I see two under 50, and a handful of others around 200 lines.

In the future, if the generator learns to parse schema files, we could instead produce Go files alongside them or following their name. For example, `foo.proto` produces `foo.pb.go` with `protoc-gen-go`, and we could similarly use a standard like `ipld_gen_${name}.go`.

/cc @warpfork @aschmahmann

**Comment by warpfork** [comment]
- I definitely agree with getting rid of the linear number of files in the number of types.  That's just silly.  It's a first-draft hangover and indefensible.  :+1: 
- A CLI tool which takes hints for output filenames from the schema file it reads sounds perfectly reasonable. :+1: 
- I'm indifferent, maybe softly negative, to combining the type table and the minima files together with the rest of the gen output. :shrug: 
- There's probably a lot of different ways we could satisfy "output file count should not be linear in count of types" while still having some files split for readability?  I'd like to investigate some. :mag: 

One idea (I've had comments outlining this on my git stash for a while, but not implemented):

- a file for the data structures
    - this is the file a go developer is most likely to actually _look_ at, to see how things got laid out in memory.
    - maybe this file is even worth generating some (brief) comments in (by default; toggleable with configuration).
- a file for all the boilerplate associated with each type
    - the node contract methods
    - the representation type and representation methods
    - any native/"speciated" methods for the type
        - (in _some_ cases it might be worth putting these in the data structure file -- for example structs with stringjoin representations; it's nice to see it when there's such a distinctive behavior that isn't directly implied by the go structure.  maybe comments would be enough, alternatively.)
    - the maybeT for the type
    - the assemblers
    - the representation assemblers
- a file for the type table
    - ... or, this just goes at the top of the file full of datastructures.  (that would be in keeping with the theme of "that's the file to look at".)
- a file for the minima
    - just because it's basically static

(I'm not entirely attached to this; it's just an option.)

To actually do this:

- Most of the code to tweak is in the ballpark of https://github.com/ipld/go-ipld-prime/blob/f9787456effa3d5dfc4c4a55d4448e1ab6eb9558/schema/gen/go/generate.go#L20 .  We could actually get everything all in one file without changing anything outside of that method, I believe.
    - We might have to chase inside the `EmitEntireType` and make alternative arrangements of the calls inside that if we want to split files in other ways.
- One big note: the whole codegenerator is currently **blind** to imports.  There are some ways we might rearrange files which would not be possible without making the generator aware of imports that a file will need.
    - I propose evading this in a very blunt way: for every possible case where an import might be needed conditionally, we instead give that thing a local alias in the package, and put those *unconditionally* in the minima file.  This keeps all the interesting imports in that file, and simplifies the rest of generation as a result.

**Comment by mvdan** [comment]
I honestly don't mind 1 or 3 files as long as it's a fixed small number, but if we do multiple files then the "name them after the input file" rule starts breaking. For an input `foo.schema`, instead of just producing `ipld_gen_foo.go`, we would presumably also start producing something like `ipld_gen_foo_type_table.go`, though that would then conflict with `foo_type_table.schema` in the same directory. Similarly, just `type_table.go` would be a conflict between schema files.

I guess we could force one single `go generate` run per package to be able to produce files like `type_table.go` all at once, though I would prefer it if we could generate from each schema file separately.

**Comment by mvdan** [comment]
I forgot to mention - Go filenames are by convention snake_case, not camelCase. So instead of generating filenames like `typeTable.go`, we should generate `type_table.go`.

**Comment by warpfork** [comment]
(Hah!  I just tried to poke at this and remembered one of the reasons I didn't put everything in one file in earlier rounds of implementation: it's harder.  Consistency of output requires either maintaining order in the type map or sorting, and I punted, thinking self-hosting gen of schema-schema types would make that vanish in a puff of smoke.

Still going to kick this around today, but we'll see if I end up rabbit-holing into that bigger topic or just doing sorting as a shortcut.)
